File: Python logic/solver.py
from player import Player
from card import Card
from details import heirarchies, numbers, suits, gameTypes
import random


def winnerAction(winner):
    """We can create some other function other than this
    such as highlighting the winner's card background"""
    if winner != 0:
        print("Player", winner, "won.")
    else:
        print("Tie")


class game:
    def __init__(self, gameType):
        """gameType is a string"""
        self.gameType = gameType
        self.cardsDealt = gameTypes[gameType]
        self.P1 = Player()
        self.P2 = Player()

    def setCard(self, player_id, suit, num):
        """appends card object to player of given player_id
        inputs are player_id,suit code(0,1,2,3) and num(2,3,4,'a','k','q')"""
        if player_id == 1:
            if len(self.P1.cards) < self.cardsDealt:
                self.P1.cards.append(Card(suit, num))
            else:
                if len(self.P2.cards) < self.cardsDealt:
                    """We can automatically switch to player 2 and fix it"""
                    print("Enter cards for player 2")
                else:
                    """We can disable the enter button and activate the compare button"""
                    print("Initiate compare")
        elif player_id == 2:
            if len(self.P2.cards) < self.cardsDealt:
                self.P2.cards.append(Card(suit, num))
            else:
                if len(self.P1.cards) < self.cardsDealt:
                    """We can automatically switch to player 1"""
                    print("Enter cards for player 1")
                else:
                    """We can disable the enter button and activate the compare button"""
                    print("Initiate compare")
        return True

    def gameType_Classic(self):
        p1 = [self.P1.hierarchy] + self.P1.hierarchyNumber
        p2 = [self.P2.hierarchy] + self.P2.hierarchyNumber
        for i in range(len(p1)):
            if p1[i] > p2[i]:
                winnerAction(1)
                break
            elif p1[i] < p2[i]:
                winnerAction(2)
                break

    def gameTypeMethodDecider(self):
        if self.gameType == "Classic":
            self.gameType_Classic()

    def compare(self):
        if len(self.P1.cards) == self.cardsDealt and len(self.P2.cards) == self.cardsDealt:
            self.P1.hierarchy_decider()
            self.P2.hierarchy_decider()
            self.gameTypeMethodDecider()
        else:
            print("Enter More cards")

    def reset(self):
        self.P1 = Player()
        self.P2 = Player()


"""Stream Looks like this"""
# gT = "Classic"
# g = game(gT)
# g.compare()
# g.setCard(1, "spade", 2)
# g.setCard(1, "spade", "a")
# g.setCard(1, "spade", 3)
# g.setCard(2, "spade", "a")
# g.setCard(2, "spade", "k")
# g.setCard(2, "spade", "q")
# g.compare()

"""Python Testing Code"""
testing_iterations = 10
s = [(k) for k, v in suits.items()]
n = [(k) for k, v in numbers.items()]
print(n)
for i in range(testing_iterations):
    p1, p2 = [], []
    gT = "Classic"
    g = game(gT)
    cards = []
    for j in range(3):
        t, suit, num = True, None, None
        while t:
            suit = random.choice(s)
            num = random.choice(n)
            if [suit, num] not in cards:
                cards.append([suit, num])
                break
        t = g.setCard(1, suit, num)
        p1 += [[suit, num]]
    for j in range(3):
        t, suit, num = True, None, None
        while t:
            suit = random.choice(s)
            num = random.choice(n)
            if [suit, num] not in cards:
                cards.append([suit, num])
                break
        t = g.setCard(2, suit, num)
        p2 += [[suit, num]]
    print(p1, p2)
    g.compare()

File: JS-logic/solver.js
const { Player, player1, player2 } = require("./player")
const { Card } = require("./card")
const { heirarchies, suits, numbers, gameTypes } = require("./details")

function winnerAction(winner) {
    console.log("Winner function")
    if (winner != 0)
        console.log("Player", winner, "won.")
    else
        console.log("Tie")
}

class Game {
    constructor(gameType) {
        this.gameType = gameType;
        this.cardsDealt = gameTypes[gameType];
        this.P1 = player1
        this.P2 = player2
    }

    setCard(player_id, suit, num) {
        // appends card object to player of given player_id
        // inputs are player_id, suit code(0, 1, 2, 3) and num(2, 3, 4, 'a', 'k', 'q')
        if (player_id === 1) {
            if (this.P1.cards.length < this.cardsDealt) {
                let c1 = new Card(suit, num);
                console.log(c1)
                console.log("P1 Cards", this.P1.cards);
                this.P1.cards.push(c1)
                console.log("P1 Cards", this.P1.cards);
            }
            else
                if (this.P2.cards.length < this.cardsDealt)
                    //We can automatically switch to player 2
                    console.log("Enter cards for player 2")
            // else
            //We can disable the enter button and activate the compare button
            // console.log("Initiate compare")
        }
        else if (player_id === 2) {
            if (this.P2.cards.length < this.cardsDealt) {
                let c1 = new Card(suit, num);
                console.log(c1)
                console.log("P2 Cards", this.P2.cards);
                this.P1.cards.push(c1)
                console.log("P2 Cards", this.P2.cards);
            }
            else
                if (this.P1.cards.length < this.cardsDealt)
                    // We can automatically switch to player 1
                    console.log("Enter cards for player 1")
            // else
            //We can disable the enter button and activate the compare button
            // console.log("Initiate compare")
        }
        return true
    }

    gameType_Classic() {
        console.log(this.P1)
        let p1 = [];
        p1.push(this.P1.hierarchy, this.P1.hierarchyNumber);
        let p2 = [];
        p2.push(this.P2.hierarchy, this.P2.hierarchyNumber);
        for (var i; i < p1.length; i++) {
            if (p1[i] > p2[i]) {
                winnerAction(1);
                return
            }
            else if (p1[i] < p2[i]) {
                winnerAction(2);
                return
            }
        }
    }

    gameTypeMethodDecider() {
        if (this.gameType === "Classic")
            this.gameType_Classic();
    }

    compare() {
        if (this.P1.cards.length === this.cardsDealt && this.P2.cards.length === this.cardsDealt) {
            this.P1.hierarchy_decider();
            this.P2.hierarchy_decider();
            this.gameTypeMethodDecider();
        }
        else
            console.log("Enter More cards")
    }
}

let testing_iterations = 2
let s = [];
for (k in suits) {
    // console.log(k)
    s.push(suits[k]);
}
let n = [];
for (k in numbers) {
    // console.log(k)
    n.push(numbers[k]);
}
for (let i = 0; i < testing_iterations; i++) {
    let p1 = [];
    let p2 = [];
    let gT = "Classic"
    let g = new Game(gT)
    let cards = [];
    for (var j = 0; j < 3; j++) {
        let t = true
        let suit = null
        let num = null
        while (t) {
            suit = s[Math.floor(Math.random() * s.length)]
            num = n[Math.floor(Math.random() * n.length)]
            if (!cards.includes([suit, num])) {
                cards.push([suit, num])
                break
            }
            console.log(1)
        }
        t = g.setCard(1, suit, num)
        p1.push([suit, num])
    }
    for (var j = 0; j < 3; j++) {
        let t = true
        let suit = null
        let num = null
        while (t) {
            suit = s[Math.floor(Math.random() * s.length)]
            num = n[Math.floor(Math.random() * n.length)]
            if (!cards.includes([suit, num])) {
                cards.push([suit, num])
                break
            }
            console.log(2)
        }
        t = g.setCard(2, suit, num)
        p2.push([suit, num])
    }
    console.log("Cards of Player 1:", p1);
    console.log("Cards of Player 2:", p2);
    g.compare()
}
